homeassistant/components/rfxtrx/__init__.py

```diff
@@ -248,29 +243,32 @@ def handle_receive(event):
 
         _LOGGER.debug("Receive RFXCOM event: %s", event_data)
 
-        data_bits = get_device_data_bits(event.device, device_bits)
+        data_bits = get_device_data_bits(event.device, devices)
         device_id = get_device_id(event.device, data_bits=data_bits)
 
         # Callback to HA registered components.
         hass.helpers.dispatcher.dispatcher_send(SIGNAL_EVENT, event, device_id)
 
         # Signal event to any other listeners
-        if device_id in device_events:
+        fire_event = devices.get(device_id, {}).get(CONF_FIRE_EVENT)
+        if fire_event:
             hass.bus.fire(EVENT_RFXTRX_EVENT, event_data)
 
-        if config[CONF_AUTOMATIC_ADD]:
-            if device_id not in devices:
-                data = entry.data
-                data[CONF_DEVICES][event_data["data"]] = device_id
-
-                asyncio.run_coroutine_threadsafe(
-                    async_add_device(hass, entry, data), hass.loop
-                )
+    @callback
+    def async_add_device_to_config(hass, event, device_id):
+        data = entry.data.copy()
+        event_code = binascii.hexlify(event.data).decode("ASCII")
+        data[CONF_DEVICES][event_code] = device_id
+        hass.config_entries.async_update_entry(entry=entry, data=data)
 
-                devices.add(device_id)
+    @callback
+    def device_update(event, device_id):
+        if device_id not in devices:
+            hass.add_job(async_add_device_to_config, hass, event, device_id)
```

Comment:
Since now we're already in async context we don't need to use `hass.add_job` anymore.

We can inline `async_add_device_to_config` inside `device_update`.

Reply:
Are we really in async context? The `device_update` is still `callback` right?
The following is not working for me, complaining about not awaiting `async_add_device_to_config`

```
async def async_add_device_to_config(hass, event, device_id):
        data = entry.data.copy()
        event_code = binascii.hexlify(event.data).decode("ASCII")
        data[CONF_DEVICES][event_code] = device_id
        await hass.config_entries.async_update_entry(entry=entry, data=data)

    @callback
    def device_update(event, device_id):
        if device_id not in devices:
            async_add_device_to_config(hass, event, device_id)
            devices[device_id] = {}
```

Reply:
Or do you mean making the `async_add_device_to_config` callback again? Like this:

```
@callback
    def async_add_device_to_config(hass, event, device_id):
        data = entry.data.copy()
        event_code = binascii.hexlify(event.data).decode("ASCII")
        data[CONF_DEVICES][event_code] = device_id
        hass.config_entries.async_update_entry(entry=entry, data=data)

    @callback
    def device_update(event, device_id):
        if device_id not in devices:
            async_add_device_to_config(hass, event, device_id)
            devices[device_id] = {}
```

Reply:
Forget about those two, I get the logic now.

Reply:
we are in async context since you add a callback function to (async_)dispatcher_connect. When that is called, we are in event loop then.

Reply:
So yes, the device_update function is no longer needed.

Reply:
Yeah, don't know why I kept it, should be fine now